### DISCUSSION TITLE
Webauthn not working on Desktop Clients

### DISCUSSION THREAD

**Discussion by bokkabonga** [dani-garcia/vaultwarden#2629] [Q&A] [answered]
### Subject of the issue
Using the Desktop Clients Webauthn gets stuck and only displays "loading". I tried to add "file://*" as described in [this](https://github.com/dani-garcia/vaultwarden/discussions/2111#discussioncomment-1688737) discussion.

### Deployment environment

* vaultwarden version: 1.25.1

* Clients used: Desktop Client (Linux, Windows)

* Reverse proxy and version: Apache

* MySQL/MariaDB or PostgreSQL version: MariaDB 10.4

Support String:

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.25.1-f1a67663
* Web-vault version: v2022.6.2
* Running within Docker: true (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: MySQL
* Database version: 10.4.20-MariaDB-1:10.4.20+maria~focal
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** DOMAIN, SENDS_ALLOWED, HIBP_API_KEY, USER_ATTACHMENT_LIMIT, ORG_ATTACHMENT_LIMIT, TRASH_AUTO_DELETE_DAYS, SIGNUPS_ALLOWED, SIGNUPS_VERIFY, SIGNUPS_VERIFY_RESEND_TIME, SIGNUPS_VERIFY_RESEND_LIMIT, SIGNUPS_DOMAINS_WHITELIST, INVITATIONS_ALLOWED, PASSWORD_HINTS_ALLOWED, ADMIN_TOKEN, INVITATION_ORG_NAME, DISABLE_2FA_REMEMBER, YUBICO_CLIENT_ID, YUBICO_SECRET_KEY, SMTP_HOST, SMTP_PORT, SMTP_FROM, EMAIL_TOKEN_SIZE, EMAIL_EXPIRATION_TIME

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_token": "***",
  "allowed_iframe_ancestors": "****://*",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "*****://****:*********************************@**/*********",
  "db_connection_retries": 10,
  "disable_2fa_remember": true,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*********.***.**",
  "domain_origin": "*****://*********.***.**",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 5 * * * *",
  "emergency_request_timeout_schedule": "0 5 * * * *",
  "enable_db_wal": false,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": "***",
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_org_name": "Bitwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": "/data/bitwarden.log",
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": 20480,
  "org_creation_users": "",
  "password_hints_allowed": false,
  "password_iterations": 100000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "***.**",
  "signups_verify": true,
  "signups_verify_resend_limit": 3,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_explicit_tls": null,
  "smtp_from": "*********@***.**",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "**********.***.**",
  "smtp_password": null,
  "smtp_port": 25,
  "smtp_security": "off",
  "smtp_ssl": false,
  "smtp_timeout": 15,
  "smtp_username": null,
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": 60,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_syslog": false,
  "user_attachment_limit": 5121,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": true,
  "websocket_port": 3012,
  "yubico_client_id": "61202",
  "yubico_secret_key": "***",
  "yubico_server": null
}
```
</details>

### Troubleshooting data
Everything i could find in the Logfile:
`[2022-07-19 11:51:27.200][request][INFO] POST /api/accounts/prelogin
[2022-07-19 11:51:27.204][response][INFO] (prelogin) POST /api/accounts/prelogin => 200 OK
[2022-07-19 11:51:27.284][request][INFO] POST /identity/connect/token
[2022-07-19 11:51:27.345][error][ERROR] 2FA token not provided
[2022-07-19 11:51:27.345][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
`

**Comment by BlackDex** [comment, marked as answer]
I just tested this my self with Windows Hello and SoftU2F on Linux Both are working as it should.
Also adding the `file://` shouldn't be needed anymore since those headers are not sent for the connectors anymore.

Also, the Linux Desktop client does not support WebAuthn at all, so that would be normal that it doesn't work.

That _error_ 400 you get is normal, that indicates the clients that it needs MFA/2FA to continue.
I suggest to open the `Developer Console` in the Desktop client by either pressing F12 or go to `View` > `Toggle developer tools`.
Refresh the page and then check either the Console or Network tab for issues with loading the needed pages.

It could be that your reverse proxy is filtering some headers, or is adding some of there own which could cause issues.

**Comment by bokkabonga** [comment]
Got it working again by removing the header filters in the apache config. I now get some other errors in the Developer Console, but as far as i can tell everything works fine. Whats strange is, that all those are not activated in the config of the apache. Maybe ist something with rocket itself?

Are those errors something to worry about or can they be ignored?

> Error with Permissions-Policy header: Unrecognized feature: 'ambient-light-sensor'.
Error with Permissions-Policy header: Unrecognized feature: 'battery'.
Error with Permissions-Policy header: Unrecognized feature: 'execution-while-not-rendered'.
Error with Permissions-Policy header: Unrecognized feature: 'execution-while-out-of-viewport'.
Error with Permissions-Policy header: Unrecognized feature: 'keyboard-map'.
Error with Permissions-Policy header: Unrecognized feature: 'web-share'.

**Reply by BlackDex** [reply]
Those are set by Vaultwarden. Some clients do not support those features, but others do, like mobile-phones.
This prevents those clients from accessing those API's. Nothing to worry about.
